### ISSUE TITLE
zramctl: output columns for incompressible pages

### ISSUE THREAD

**Issue by polarathene** [karelzak/util-linux#1347]
https://github.com/karelzak/util-linux/blob/1e881378d6674abaae5a929f7deff39bb3405c3b/sys-utils/zramctl.c#L74-L87

I noticed that `mm_stat` data for incompressible pages (`huge_pages`) is missing, is it worth adding for output as well?

Additionally in the 5.11 kernel support was added for [incompressible pages since zram device initialized](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=194e28da1a0279ef6a106a5b621fd79c410432ef) (`huge_pages_since`), not just currently stored. Both are mentioned in the [kernel docs for zram](https://www.kernel.org/doc/html/latest/admin-guide/blockdev/zram.html).

I think `huge_pages` stat for showing the currently stored incompressible pages can be a useful complimentary stat beside the "ZERO-PAGES" column (`same_pages` in mm_stat output).

**Comment by karelzak** [comment]
Good point. It would be nice to sync the code in zramctl with the current kernel -- it means add huge_pages and huge_pages_since. 

The question is how to implement it in backwardly compatible way, there's 
```c
/* make sure kernel provides mm_stat as expected */
if (strv_length(z->mm_stat) < ARRAY_SIZE(mm_stat_names))
   ...
```
in zramctl.c.  I guess we need to make it usable with old kernels (fewer fields in the mm_stat) as well as with the latest 5.11. Maybe remove the check at all and do not use directly z->mm_stat[idx] when asking for a field, but cover it with some protection, something like:
```c
z->mm_nfields = strv_length(z->mm_stat);
...
if (idx >= z->mm_nfields)
    return xstrdup("0");
else
    return xstrdup(z->mm_stat[idx]);
```
... or so.

**Comment by polarathene** [comment]
AFAIK, any new stats are just appended? So you should be able to check the length(eg by whitespace delimiters?) and then pass that value to a function to know how much to parse/return? A loop should work for that?

I haven't looked at the source too much, not an area I'm familiar enough with. Your code snippet makes enough sense to me for the logic though :+1:

**Comment by karelzak** [comment]
I *hope* the kernel maintains backward compatibility here and it only appends to the file ;-) I cannot imagine another solution for files with unnamed values.

**Comment by OmegaSquad82** [comment]
As [recompression](https://docs.kernel.org/admin-guide/blockdev/zram.html#recompression ) has been a thing within the kernel for a while and tools like systemd-zram-generator [recently](https://github.com/systemd/zram-generator/issues/178) implemented support to elegantly configure it, it would be nice to display above mentioned information.

Perhaps even add to `--output-all` some statistics on [memory tracking](/sys/kernel/debug/zram/zram0/block_state), if it is available, or would this be more suitable for an additional feature request?

**Comment by karelzak** [comment]
There is more to add for the recompress.

It would be nice to have a command line option for setup.
```
 --recompress <setting>
```
where the setting is a list of variables 
```
 algorithm=<name>
 priority=<number>
 threshold=<number>
 type=<huge|idle|huge_idle>
 max_pages=<number>
```
And we can also add output columns RECOMP-{ALGO,TYPE,...} for all of these settings
